lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -388,201 +378,529 @@ export class UserAgentApplication {
+      
+      // Open the popup window
+      var popUpWindow = this.openWindow("about:blank", "_blank", 1, this, resolve, reject);
+      if (!popUpWindow) {
+        // TODO: we should be rejecting with an error here
+        return;
+      }
+
+      acquireTokenAuthority.ResolveEndpointsAsync().then(() => {
+        // On fullfillment
+        if (Utils.compareObjects(userObject, this.getUser())) {
+          if (scopes.indexOf(this.clientId) > -1) {
+            authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.id_token, this.getRedirectUri(), this._state);
+          }
+          else {
+            authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.token, this.getRedirectUri(), this._state);
+          }
+        } else {
+          authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.id_token_token, this.getRedirectUri(), this._state);
+        }
+
+        // Cache nonce
+        // TODO: why is storeAuthStateInCookie not passed here?
+        this._cacheStorage.setItem(Constants.nonceIdToken, authenticationRequest.nonce);
+        authenticationRequest.state = authenticationRequest.state;
+
+        // Cache acquireTokenUserKey
+        var acquireTokenUserKey;
+        if (userObject) {
+          acquireTokenUserKey = Constants.acquireTokenUser + Constants.resourceDelimeter + userObject.userIdentifier + Constants.resourceDelimeter + authenticationRequest.state;
+        }
+        else {
+          acquireTokenUserKey = Constants.acquireTokenUser + Constants.resourceDelimeter  + Constants.no_user + Constants.resourceDelimeter + authenticationRequest.state;
+        }
+        this._cacheStorage.setItem(acquireTokenUserKey, JSON.stringify(userObject));
+
+        // Cache authorityKey
+        const authorityKey = Constants.authority + Constants.resourceDelimeter + authenticationRequest.state;
+        this._cacheStorage.setItem(authorityKey, acquireTokenAuthority.CanonicalAuthority, this.storeAuthStateInCookie);
+
+        // Set the extraQueryParameters
+        if (extraQueryParameters) {
+          authenticationRequest.extraQueryParameters = extraQueryParameters;
+        }
+
+        // Construct the urlNavigate
+        let urlNavigate = authenticationRequest.createNavigateUrl(scopes) + Constants.response_mode_fragment;
+        urlNavigate = this.addHintParameters(urlNavigate, userObject);
+        window.renewStates.push(authenticationRequest.state);
+        window.requestType = Constants.renewToken;
+        this.registerCallback(authenticationRequest.state, scope, resolve, reject);
+
+        // open popup window to urlNavigate
+        if (popUpWindow) {
+          popUpWindow.location.href = urlNavigate;
+        }
+
+      }, () => {
+        // On rejection
```

Comment:
some of this looks duplicated from above

Reply:
Duplicated code exists currently, we are working to reduce this in future PRs.